server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.server.manager.state;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.accumulo.core.client.ConditionalWriter;
+import org.apache.accumulo.core.clientImpl.ClientContext;
+import org.apache.accumulo.core.dataImpl.KeyExtent;
+import org.apache.accumulo.core.metadata.TServerInstance;
+import org.apache.accumulo.core.metadata.TabletLocationState;
+import org.apache.accumulo.core.metadata.schema.Ample;
+import org.apache.accumulo.core.metadata.schema.TabletMetadata;
+import org.apache.accumulo.core.tabletserver.log.LogEntry;
+import org.apache.accumulo.server.util.ManagerMetadataUtil;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.base.Preconditions;
+
+public abstract class AbstractTabletStateStore implements TabletStateStore {
+
+  private final ClientContext context;
+  private final Ample ample;
+
+  protected AbstractTabletStateStore(ClientContext context) {
+    this.context = context;
+    this.ample = context.getAmple();
+  }
+
+  @Override
+  public void setLocations(Collection<Assignment> assignments) throws DistributedStoreException {
+    try (var tabletsMutator = ample.conditionallyMutateTablets()) {
+      for (Assignment assignment : assignments) {
+        var conditionalMutator = tabletsMutator.mutateTablet(assignment.tablet)
+            .requireAbsentOperation()
+            .requireLocation(TabletMetadata.Location.future(assignment.server))
+            .putLocation(TabletMetadata.Location.current(assignment.server))
+            .deleteLocation(TabletMetadata.Location.future(assignment.server)).deleteSuspension();
```

Comment:
It appears that `ConditionalWriter.convertConditions` will sort the Conditions based on CF, CQ, Visibility, and then timestamp. Given that `requireLocation`, `putLocation`, and `deleteLocation` all operate on the same CF and CQ with a timestamp of 0L, do we know in which order they are applied?

Reply:
The requireLocation call will create a condition on the conditional mutation.  All conditions on a conditional mutation are checked first before any changes are made.  So a check is done to see if the future location is as expected and if it is then the changes to delete the future and set the current are made.   If the check of the current future location fails then no changes are made.

Reply:
So what happens on the server is

 1. it gets a lock on the row
 2. it checks the cf=future and cq=session match what was specified in requireLocation(TabletMetadata.Location.future(assignment.server))
 3. If the above check was succesful it deletes the future col and sets the current col
 4. releases the row lock

Reply:
Sorry for asking so many questions about this, I'm not familiar with ConditionalMutations. In the code above there is an implied order in the conditions, but it appears that they are re-sorted on the client side before sending to the server [here](https://github.com/apache/accumulo/blob/ba472d6e24daa8f0014a22cabace3061f5d46413/core/src/main/java/org/apache/accumulo/core/clientImpl/ConditionalWriterImpl.java#L777) which is called as part of the write process in `ConditionalWriterImpl.sendToServer`. 

Are all of the conditions checked before any changes are applied and therefore the order is irrelevant?

Reply:
> Are all of the conditions checked before any changes are applied and therefore the order is irrelevant?

That's correct.  The code that sorts the conditions just does that to make them more efficient to check.

Reply:
ok, thanks. I think that alleviates my concerns.